src/dask_nested/core.py

```diff
@@ -0,0 +1,390 @@
+    ) -> NestedFrame:
+        """Returns an EnsembleFrame constructed from a TapeFrame.
+
+        Parameters
+        ----------
+        data: `TapeFrame`
+            Frame containing the underlying data fro the EnsembleFram
+        npartitions: `int`, optional
+            The number of partitions of the index to create. Note that depending on
+            the size and index of the dataframe, the output may have fewer
+            partitions than requested.
+        chunksize: `int`, optional
+            Size of the individual chunks of data in non-parallel objects that make up Dask frames.
+        sort: `bool`, optional
+            Whether to sort the frame by a default index.
+        label: `str`, optional
+            The label used to by the Ensemble to identify the frame.
+        ensemble: `tape.Ensemble`, optional
+            A link to the Ensemble object that owns this frame.
+
+        Returns
+        ----------
+        result: `tape.EnsembleFrame`
+            The constructed EnsembleFrame object.
+        """
+        result = dd.from_pandas(data, npartitions=npartitions, chunksize=chunksize, sort=sort)
+        return NestedFrame.from_dask_dataframe(result)
+
+    @classmethod
+    def from_dask_dataframe(cls, df) -> NestedFrame:
+        """Converts a Dask Dataframe to a Dask-Nested NestedFrame
+
+        Parameters
+        ----------
+        df:
+            A Dask Dataframe to convert
+
+        Returns
+        -------
+        `dask_nested.NestedFrame`
+        """
+        return df.map_partitions(npd.NestedFrame)
+
+    def compute(self, **kwargs):
+        """Compute this Dask collection, returning the underlying dataframe or series."""
+        return npd.NestedFrame(super().compute(**kwargs))
+
+    @property
+    def all_columns(self) -> dict:
+        """returns a dictionary of columns for each base/nested dataframe"""
+        all_columns = {"base": self.columns}
+        for column in self.columns:
+            if isinstance(self[column].dtype, NestedDtype):
+                nest_cols = list(self.dtypes[column].fields.keys())
+                all_columns[column] = nest_cols
+        return all_columns
+
+    @property
+    def nested_columns(self) -> list:
+        """retrieves the base column names for all nested dataframes"""
+        nest_cols = []
+        for column in self.columns:
+            if isinstance(self[column].dtype, NestedDtype):
+                nest_cols.append(column)
+        return nest_cols
+
+    def add_nested(self, nested, name) -> NestedFrame:  # type: ignore[name-defined] # noqa: F821
+        """Packs a dataframe into a nested column
+
+        Parameters
+        ----------
+        nested:
+            A flat dataframe to pack into a nested column
+        name:
+            The name given to the nested column
+
+        Returns
+        -------
+        `dask_nested.NestedFrame`
+        """
+        nested = nested.map_partitions(lambda x: pack_flat(x)).rename(name)
```

Comment:
What happens if we try adding a flat dask dataframe? How would pack_flat handle it?

Reply:
It seems to fail if given a dask dataframe instead of a nested-dask nestedframe: `*** ValueError: max() arg is an empty sequence`. I can open a ticket to support this.

Reply:
Yeah just a ticket should be fine for now.

But this seems like a blocker for out-of-memory datasets at the moment which is unfortunate

Reply:
I think out-of-memory datasets are still feasible. It seems to work fine with a flat Nested-Dask NestedFrame object so the main hiccup is just that we will need to convert any input dask dataframe to a NestedFrame before calling add_nested

Reply:
We should think about a short-cut for index-sorted datasets, it should work for out-of-memory datasets